vertexai/genai/tokenizer/corpora_test.go

```diff
@@ -0,0 +1,274 @@
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package tokenizer
+
+import (
+	"archive/zip"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"regexp"
+	"strings"
+	"testing"
+
+	"cloud.google.com/go/vertexai/genai"
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/encoding/charmap"
+	"golang.org/x/text/encoding/japanese"
+	"golang.org/x/text/encoding/simplifiedchinese"
+	"golang.org/x/text/transform"
+)
+
+// FileInfo holds the name and content  of a file in the zip archive
+type CorporaInfo struct {
+	Name    string
+	Content []byte
+}
+
+// fileGenerator is a generator function that returns a channel to iterate over files in the zip archive
+func CorporaGenerator(url string) (<-chan CorporaInfo, <-chan error) {
+	out := make(chan CorporaInfo)
+	errCh := make(chan error, 1)
+
+	go func() {
+		defer close(out)
+		defer close(errCh)
+
+		// Download the zip file
+		resp, err := http.Get(url)
+		if err != nil {
+			errCh <- fmt.Errorf("error downloading file: %v", err)
+			return
+		}
+		defer resp.Body.Close()
+
+		// Read the content of the response body
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			errCh <- fmt.Errorf("error reading response body: %v", err)
+			return
+		}
+
+		// Create a zip reader from the downloaded content
+		zipReader, err := zip.NewReader(bytes.NewReader(body), int64(len(body)))
+		if err != nil {
+			errCh <- fmt.Errorf("error creating zip reader: %v", err)
+			return
+		}
+
+		// Iterate over each file in the zip archive
+		for _, file := range zipReader.File {
+			fileReader, err := file.Open()
+			if err != nil {
+				errCh <- fmt.Errorf("error opening file: %v", err)
+				continue
+			}
+
+			// Check if the file is a text file
+			if !file.FileInfo().IsDir() && file.FileInfo().Mode().IsRegular() {
+				content, err := io.ReadAll(fileReader)
+				if err != nil {
+					errCh <- fmt.Errorf("error reading file: %v", err)
+					fileReader.Close()
+					continue
+				}
+				fileReader.Close()
+
+				out <- CorporaInfo{
+					Name:    file.Name[len("udhr/"):],
+					Content: content,
+				}
+			} else {
+				fileReader.Close()
+			}
+		}
+	}()
+
+	return out, errCh
+}
+
+// UdhrCorpusReader struct to hold encoding patterns and skip set
+type UdhrCorpusReader struct {
+	Encodings []EncodingPattern
+	Skip      map[string]bool
+}
+
+// EncodingPattern struct to hold regex pattern and corresponding encoding
+type EncodingPattern struct {
+	Pattern  *regexp.Regexp
+	Encoding encoding.Encoding
+}
+
+// NewUdhrCorpusReader initializes a new UdhrCorpusReader with encoding patterns and skip set
+func NewUdhrCorpusReader() *UdhrCorpusReader {
```

Comment:
Nit: `Reader` has a certain meaning in Go (https://pkg.go.dev/io#Reader), so this naming can be confusing. AFAICS this type isn't a reader, it just describes the corpus?

Reply:
Removed the Reader suffix